packages/contracts/contracts/registry/ERC165ContractRegistry.sol

```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.10;
+
+import "../core/component/Component.sol";
+
+/// @title An ERC165-based registry for contracts
+/// @author Aragon Association - 2022
+/// @notice This contract allows to register contracts
+contract ERC165ContractRegistry is Component {
+    bytes4 internal constant REGISTRY_INTERFACE_ID = this.register.selector;
+    bytes32 public constant REGISTER_ROLE = keccak256("REGISTER_ROLE");
+
+    bytes4 public contractInterfaceId;
+
+    mapping(address => bool) public registrees;
+
+    /// @notice Thrown if the contract is already registered
+    /// @param registrant The address of the contract to be registered
+    error ContractAlreadyRegistered(address registrant);
+
+    /// @notice Thrown if the contract does not support the required interface
+    /// @param registrant The address of the contract to be registered
+    error ContractInterfaceInvalid(address registrant);
+
+    /// @notice Initializes the component
+    /// @dev This is required for the UUPS upgradability pattern
+    /// @param _managingDao The interface of the DAO managing the components permissions
+    /// @param _contractInterfaceId The ERC165 interface id of the contracts to be registered
+    function initialize(IDAO _managingDao, bytes4 _contractInterfaceId) public initializer {
+        _registerStandard(REGISTRY_INTERFACE_ID);
+
+        __Component_init(_managingDao);
+
+        contractInterfaceId = _contractInterfaceId;
+    }
+
+    /// @notice Registers a contract by its address
+    /// @param registrant The address of the contract to be registered
+    function register(address registrant) external auth(REGISTER_ROLE) {
+        _register(registrant);
+    }
+
+    function _register(address registrant) internal {
+        if (!AdaptiveERC165(registrant).supportsInterface(contractInterfaceId))
+            revert ContractInterfaceInvalid(registrant);
+
+        if (registrees[registrant]) revert ContractAlreadyRegistered({registrant: registrant});
+
+        registrees[registrant] = true;
```

Comment:
move this to the beginning of the function to because `supportsInterface` is an external call, isn't it?

Reply:
What's the point in tracking that a contract was registered?

Without any event emitted or any way to read things, what is the goal of this contract?

Reply:
With `mapping(address => bool) public registrees` you can check if the contract is registered.

In this design, events are expected to be emitted in the context, where `function register(address registrant) external auth(REGISTER_ROLE)` is used. 
This way, we don't need to create multiple specialized versions of `ERC165Registry` only differing by the event that they emit.

Reply:
> you can check if the contract is registered
Right, but does this thing alone justify the existence of an address => bool registry? What's the use case? What ERC165 feature are we getting in either of these 2 contract versions? :)

> events are expected to be emitted in the context, where function register(address registrant) external auth(REGISTER_ROLE) is used
Ok, so this assumes a contract inheriting from it, adding such event? If so, shouldn't this contract be left as abstract?